proposals/secrets-support.md

```diff
@@ -0,0 +1,85 @@
+# First class secrets support in devcontainers CLI
+
+## What are secrets
+Secrets are variables that hold sensitive values and needs to be handled securely at all times (API keys, passwords etc.). Users can change a secret's value at anytime, and the devcontainers CLI should support dynamically changing secrets without having to rebuild the container.
+
+## Goal
+
+Support Secrets as a first class feature in devcontainers CLI.
+
+This feature should comprise of the ability to securely pass in the secrets in CLI commands, make them available for users to consume similar to remoteEnv and containerEnv, and be processed and handled securely at all times.
+This functionality will allow consumers to be able to use secrets more predictably and securely with devcontainers CLI.
+
+## Motivation
+
+Today Codespaces pass secret variables as `remoteEnv` since there are no other mechanisms to pass them explicitly as secrets. The devcontainers CLI do not (and need not) treat `remoteEnv` as secrets. Having explicit support for secrets will help eliminating security threats such as leaking secrets, as well as in improving customer confidence.
+
+## Proposal
+
+Following are the three main functional changes to the CLI to support this feature.
+
+  1. Ability to pass secrets to commands
+  2. Apply/use secrets similar to `remoteEnv`
+  3. Securely handle secrets
+
+### Passing secrets to commands
+Secrets are not part of devcontainers specification and we do not expect users to store secrets inside devcontainer.json file.
+
+Secrets should be passed to the following devcontainers CLI commands.
+
+#### **Phase 1**
+ - up
+ - run-user-commands
+
+`up`, `run-user-commands` will use the secrets for usecases like lifecycle hooks (similar to remoteEnv). We do not plan secrets support for `build` command since we do not have a demostrated requirement, as well as to avoid potentially baking in the user secrets into the docker images.
```

Comment:
```suggestion
`up`, `run-user-commands` will use the secrets for usecases like lifecycle hooks (similar to remoteEnv). We do not plan secrets support for `build` command since we do not have a demonstrated requirement, as well as to avoid potentially baking in the user secrets into the docker images.
```